### PR TITLE
108 place an order

### PR DESCRIPTION
**Pull request by muhammad-ahmad45**

### CODE REVIEW

File: .byebug_history
Comment (ghost) [inline .byebug_history:1]: Remove this File.

File: app/controllers/line_items_controller.rb
Comment (ghost) [inline app/controllers/line_items_controller.rb:1]: Remove this Byebug

File: app/controllers/orders_controller.rb [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    else
+      render :new
+=======
+    @order = @cart.create_order(order_params)
+    if @order
+      redirect_to '/orders', method: :get
+>>>>>>> parent of 4f0e174 (Order: added place an order functionality)
+    end
+  end
+
```
Comment (ghost) [inline app/controllers/orders_controller.rb]: Fix this.

File: app/controllers/orders_controller.rb [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    else
+      render :index, status: :unprocessable_entity
+    end
+  end
+
+  def show
+<<<<<<< HEAD
+    @order = Order.find(params[:id])
+=======
+>>>>>>> parent of 4f0e174 (Order: added place an order functionality)
```
Comment (ghost) [inline app/controllers/orders_controller.rb]: Fix this too.

File: app/javascript/cart.js [hunk 3]
```diff
@@ -58,7 +58,7 @@ $(document).on('turbo:load', function() {
   function updateQuantity(lineItemId, newQuantity) {
     lineItemId = parseInt(lineItemId); // Convert lineItemId to an integer
     $.ajax({
-      url: 'line_items/23/update_quantity',
+      url: 'line_items/52/update_quantity',
```
Comment (ghost) [inline app/javascript/cart.js]: Fix the Ajax feature.

File: app/mailers/order_mailer.rb [hunk 4]
```diff
@@ -0,0 +1,6 @@
+class OrderMailer < ApplicationMailer
+    def confirmation_email(order)
+        @order = order
+        mail(to: @order.user.email, subject: 'Order Confirmation')
+    end
+end
```
Comment (ghost) [inline app/mailers/order_mailer.rb:6]: Use 2 spaces for Indentation

File: app/views/layouts/_header.html.erb [hunk 5]
```diff
@@ -6,14 +6,22 @@
...
+            <a class="nav-link active" aria-current="page" href="/users">Users</a>
+          </li>
+          <li class="nav-item">
+            <a class="nav-link active" aria-current="page" href="/orders">All Orders</a>
+          </li>
+        <% end %>
         <li class="nav-item">
           <a class="nav-link" href="/products">Products</a>
         </li>
+        <%if current_user && (current_user.if_admin? || current_user.has_staff?) %>
```
Comment (ghost) [inline app/views/layouts/_header.html.erb]: Use admin? instead of if_admin? for the method name

File: app/views/orders/_form.html.erb [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        </div>
+        
+        <div>
+          <%= form.hidden_field :status, value: 'dispatched' %>
+        </div>
+
+        <div class="text-center">
+          <%= form.submit 'Place Order', class: 'btn btn-primary' %>
+        </div>
+=======
```
Comment (ghost) [inline app/views/orders/_form.html.erb]: Fix these code issues.

File: app/views/orders/_form.html.erb [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    <div>
+      <%= form.label :address %><br>
+      <%= form.text_field :address %>
+      <% order.errors.full_messages_for(:address).each do |message| %>
+        <div><%= message %></div>
+>>>>>>> parent of 4f0e174 (Order: added place an order functionality)
+      <% end %>
+    </div>
+    
+    <div>
```
Comment (ghost) [inline app/views/orders/_form.html.erb]: Fix this too

File: app/views/orders/edit.html.erb
Comment (ghost) [inline app/views/orders/edit.html.erb:1]: Either use this as the edit view or delete this.
Why is this here?

File: app/controllers/orders_controller.rb [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    @orders = Order.all
+  end
+  
+  def new
+    @order = Order.new
+  end
+
+  def create
+    @cart = current_user.cart
+    @order = Order.new(order_params)
```
Comment (AyishmAziz123) [inline app/controllers/orders_controller.rb]: why are we not using @cart.build_order ?

File: app/controllers/orders_controller.rb [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    @order.cart = @cart
+    @order.user = current_user
+    if @order.save
+      OrderMailer.confirmation_email(@order).deliver_now
```
Comment (AyishmAziz123) [inline app/controllers/orders_controller.rb]: Move this to after_action

File: app/controllers/orders_controller.rb [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+      redirect_to order_path(@order)
+      @cart.line_items.destroy_all
```
Comment (AyishmAziz123) [inline app/controllers/orders_controller.rb]: are we not using dependent: :destroy for line_items association ?

File: app/controllers/orders_controller.rb [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+    else
+      render :new
+    end
+  end
+
+  def update
+    @order = Order.find(params[:id])
```
Comment (AyishmAziz123) [inline app/controllers/orders_controller.rb]: Move this to a before_action

File: app/controllers/orders_controller.rb [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    if @order.update(update_params)
+      redirect_to orders_path, method: :get, notice: "Status of Order #{@order.id} is updated successfully."
```
Comment (AyishmAziz123) [inline app/controllers/orders_controller.rb]: Why have we used method: :get

File: app/controllers/orders_controller.rb [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+    else
+      render :index, status: :unprocessable_entity
+    end
+  end
+
+  def show
+    @order = Order.find(params[:id])
```
Comment (AyishmAziz123) [inline app/controllers/orders_controller.rb]: Move this to before_action as well

File: app/models/user.rb [hunk 14]
```diff
@@ -7,9 +7,10 @@ class User < ApplicationRecord
   has_many :permissions, dependent: :destroy
   has_many :roles, through: :permissions , dependent: :destroy
   has_one :cart
+  has_many :orders
```
Comment (AyishmAziz123) [inline app/models/user.rb]: Why haven't we used dependent destroy here ?

File: app/models/user.rb [hunk 15]
```diff
@@ -23,8 +24,9 @@ def has_staff?
     self.roles.exists?(user_role: 'staff')
   end
 
-  def assign_default_customer 
-    self.permissions.build(role_id: Role.where(user_role: 'customer').first.id)
+  def assign_default_customer
+    customer_role = Role.find_by(user_role: 'customer')
+    self.permissions.build(role_id: customer_role&.id)
```
Comment (AyishmAziz123) [inline app/models/user.rb]: What will happen if customer_role is nil. Will it build permissions with role_id: nil ?

File: app/views/carts/show.html.erb [hunk 16]
```diff
@@ -24,9 +25,9 @@
               <strong>Unit Price: </strong>
               <%= line_item.product.price %>
             </p>
-            <% bill = line_item.product.price * line_item.quantity %>
             <p class="card-text">
               <strong>Total Price: </strong>
+              <% bill = line_item.total_price %>
```
Comment (AyishmAziz123) [inline app/views/carts/show.html.erb]: Should we need to eager load line_items for cart ?

File: app/views/carts/show.html.erb [hunk 17]
```diff
@@ -48,15 +49,19 @@
...
+<br>
 <div class="d-flex justify-content-center align-items-center mt-3">
   <%= button_to 'Remove All Items', user_cart_path(current_user, cart_id: @cart.id), method: :delete, class: "btn btn-danger", form: { 
     data: { 
-      turbo_confirm: "Are you sure you want remove all items from cart?"
+      turbo_confirm: "Are you sure you want to remove all items from the cart?"
     }
    } %>
+   
+   <%= button_to "Check Out", new_order_path(@cart), method: :get, class: "btn btn-primary" %>
```
Comment (AyishmAziz123) [inline app/views/carts/show.html.erb]: Please verify do we need to specify method: :get. If not then remove this from here

File: app/views/devise/shared/_links.html.erb [hunk 18]
```diff
@@ -1,25 +1,34 @@
...
-  <%= link_to "Sign up", new_registration_path(resource_name) %><br />
-<% end %>
+        <% if devise_mapping.registerable? && controller_name != 'registrations' %>
+          <%= link_to "Sign up", new_registration_path(resource_name) %><br />
+        <% end %>
 
-<%- if devise_mapping.recoverable? && controller_name != 'passwords' && controller_name != 'registrations' %>
-  <%= link_to "Forgot your password?", new_password_path(resource_name) %><br />
-<% end %>
+        <% if devise_mapping.recoverable? && controller_name != 'passwords' && controller_name != 'registrations' %>
```
Comment (AyishmAziz123) [inline app/views/devise/shared/_links.html.erb:13]: refactor `controller_name != 'passwords' && controller_name != 'registrations'`
to 
`%w[passwords registrations].exclude? controller_name`

Also move such conditions to helpers

File: app/views/orders/_form.html.erb [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+            <div><%= message %></div>
+          <% end %>
+        </div>
+      
+        <div>
+          <%= form.hidden_field :total_bill, value: current_user.cart.line_items.sum(:total_price) %>
+        </div>
+        
+        <div>
+          <%= form.hidden_field :status, value: 'dispatched' %>
```
Comment (AyishmAziz123) [inline app/views/orders/_form.html.erb]: Why are we sending this value from here ? Can we use default value for status in the database ?

File: app/views/orders/_form.html.erb [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+      <table class="table">
+        <thead>
+          <tr>
+            <th>Name</th>
+            <th>Quantity</th>
+            <th>Total Price(Rs)</th>
+          </tr>
+        </thead>
+        <tbody>
+          <% current_user.cart.line_items.each do |line_item| %>
```
Comment (AyishmAziz123) [inline app/views/orders/_form.html.erb:37]: check if we need to eager load associations here

File: app/views/orders/_update_status.html.erb [hunk 21]
```diff
@@ -0,0 +1,12 @@
+<%= form_for order, url: { controller: 'orders', action: 'update', id: order.id }, method: :patch do |form| %>
```
Comment (AyishmAziz123) [inline app/views/orders/_update_status.html.erb]: can we just user order_path(order) in place of `{ controller: 'orders', action: 'update', id: order.id }`

File: app/views/orders/edit.html.erb [hunk 22]
```diff
@@ -0,0 +1,2 @@
+<h1>Orders#edit</h1>
```
Comment (AyishmAziz123) [inline app/views/orders/edit.html.erb]: Are we using this file ?

File: app/views/orders/index.html.erb [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+      <th> <h5> Order Status  </h5> </th> 
+      </tr>
+    </thead>
+    <tbody>
+      <% @orders.each do |order| %>
+        <tr>
+          <td><%= order.id %></td>
+          <td><%= order.address %></td>
+          <td><%= order.total_bill %> Rs</td>
+          <td><%= order.user.id %></td>
```
Comment (AyishmAziz123) [inline app/views/orders/index.html.erb]: Check for n + 1 query here and on other places for this PR

File: test/mailers/previews/order_mailer_preview.rb [hunk 24]
```diff
@@ -0,0 +1,4 @@
+# Preview all emails at http://localhost:3000/rails/mailers/order_mailer
+class OrderMailerPreview < ActionMailer::Preview
```
Comment (AyishmAziz123) [inline test/mailers/previews/order_mailer_preview.rb]: Are we using this file ?

**Review by AyishmAziz123 — CHANGES_REQUESTED** [review]
Added some comments

**Pull request merged**